### ISSUE TITLE
ZIO layer initialization deadlock

### ISSUE THREAD

**Issue by szymonrodziewicz** [zio/zio#3965]
Hi. I'm having a problem with ZIO layer initialization. Calling `Runtime.unsafeFromLayer` causes my application to go into a deadlock (infinite await in OneShot.get) with following thread dump:
```
"qtp674483268-13" #13 prio=5 os_prio=31 tid=0x00007fa371bd5800 nid=0x7003 in Object.wait() [0x0000700009b0f000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(Native Method)
        - waiting on <0x00000006c06424c0> (a zio.internal.OneShot)
        at java.lang.Object.wait(Object.java:502)
        at zio.internal.OneShot.get(OneShot.scala:79)
        - locked <0x00000006c06424c0> (a zio.internal.OneShot)
        at zio.Runtime.unsafeRunSync(Runtime.scala:82)
        at zio.Runtime.unsafeRunSync$(Runtime.scala:77)
        at zio.Runtime$$anon$2.unsafeRunSync(Runtime.scala:226)
        at zio.Runtime.unsafeRun(Runtime.scala:57)
        at zio.Runtime.unsafeRun$(Runtime.scala:56)
        at zio.Runtime$$anon$2.unsafeRun(Runtime.scala:226)
        at zio.Runtime$.unsafeFromLayer(Runtime.scala:250)
        at *My entry point that calls Runtime.unsafeFromLayer*
```
I was unable to reproduce this error up to this point, but I'm working on it and if I manage to do it, I will post there the code as soon as possible. I thought that some of you may've got into a similar issue. This is how the dependencies in these services look like, I also initialize them in a similar manner:
```scala
  class AService
  class BService 
  class CService 
  class DService 

  val A = ZLayer.fromService((clock: Clock.Service) => new AService )
  val B = ZLayer.fromService((system: System.Service) => new BService)
  val C = ZLayer.fromServices[System.Service, AService, CService]((system: System.Service, a: AService) => new CService)
  val D = ZLayer.fromService((c: CService) => new DService)

  implicit val runtime: Runtime.Managed[Has[DService]] = Runtime.unsafeFromLayer(
    zio.ZEnv.live >+> A ++ B >+> C >>> D
  )
```
However, it's an oversimplification of my application logic and this snippet is not causing a deadlock. The services in fact are much more complex and what I'm trying to do now is find out which part of them is causing the issue. I would also be really grateful for hints on how could I possibly get closer to the cause. I also tried expanding the thread pool size.

**Comment by adamgfraser** [comment]
@szymonrodziewicz We have seen some issues related to Scala object initialization. See https://github.com/zio/zio/issues/3871. I would try defining your layers as `lazy val` instead of `val`. If this is the issue we have a fix but it is not binary compatible so we are going to try to get all the binary compatible changes in first. If that doesn't resolve the issue we can troubleshoot more.

**Comment by adamgfraser** [comment]
@szymonrodziewicz Can you check this on the latest snapshot? We believe we have resolved these issues but if you are still having issues would be good to know so we can investigate further.

**Comment by adamgfraser** [comment]
@szymonrodziewicz Closing as this appears to be fixed by #3919. If you are still having issues please feel free to reopen and we will investigate further.
